mithril-aggregator/src/services/usage_reporter.rs

```diff
@@ -0,0 +1,307 @@
+                let delta: i64 = (*value as i64) - (*last_value as i64);
+                (name.clone(), delta)
+            })
+            .filter(|(_name, value)| *value != 0)
+            .collect()
+    }
+
+    fn send_metrics(&mut self, duration: &Duration) {
+        let metrics = self.metric_service.export_metrics_map();
+        let delta = UsageReporter::metrics_delta(&self.last_metrics, &metrics);
+        let date = Utc::now();
+
+        self.last_metrics = metrics;
+
+        for (name, value) in delta {
+            let _result = self
+                .transmitter_service
+                .send_event_message::<MetricMessage>(
+                    "Metrics",
+                    &name,
+                    &MetricMessage {
+                        counter: value,
+                        duration: *duration,
+                        date,
+                    },
+                    vec![],
+                );
+        }
+    }
+
+    /// Start a loop that send event about metrics at the given time interval.
+    pub async fn run_forever(&mut self, run_interval: Duration) {
+        let mut interval = tokio::time::interval(run_interval);
+
+        loop {
+            interval.tick().await;
+            self.send_metrics(&run_interval);
+
+            info!(
+                self.logger,
+                "Metrics sent, Sleeping for {} seconds",
+                run_interval.as_secs()
+            );
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use mithril_metric::MetricCollector;
+    use tokio::sync::mpsc::UnboundedReceiver;
+
+    use super::UsageReporter;
+    use super::*;
+    use crate::test_tools::TestLogger;
+
+    fn build_usage_reporter() -> (
+        UsageReporter,
+        Arc<MetricsService>,
+        UnboundedReceiver<EventMessage>,
+    ) {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel::<EventMessage>();
+        let metric_service = Arc::new(MetricsService::new(TestLogger::stdout()).unwrap());
+        let transmitter_service = Arc::new(TransmitterService::new(tx, TestLogger::stdout()));
+        let usage_reporter = UsageReporter::new(
+            transmitter_service.clone(),
+            metric_service.clone(),
+            TestLogger::stdout(),
+        );
+        (usage_reporter, metric_service, rx)
+    }
+
+    fn received_messages(
+        rx: &mut tokio::sync::mpsc::UnboundedReceiver<EventMessage>,
+    ) -> Vec<EventMessage> {
+        let mut received_messages: Vec<EventMessage> = Vec::new();
+        while let Ok(message) = rx.try_recv() {
+            received_messages.push(message);
+        }
+        received_messages
+    }
+
+    fn extract_metric_value(message: &EventMessage) -> (String, i64) {
+        let metric_delta: MetricMessage = serde_json::from_value(message.content.clone()).unwrap();
+        (message.action.clone(), metric_delta.counter)
+    }
+
+    #[tokio::test]
+    async fn when_no_metrics_no_message_sent() {
```

Comment:
This test can be defined as synchronous, as well as the three below.
```suggestion
    #[test]
    fn when_no_metrics_no_message_sent() {
```

Reply:
Done